File: backend/repositories/paperSizeRepository.js
import db from '../config/db.js';

/**
 * Mendapatkan semua paper sizes
 * @returns {Promise<Array>} Array of paper_sizes
 */
export const getAllPaperSizes = async () => {
  const [rows] = await db.query('SELECT * FROM paper_sizes WHERE is_active = 1 ORDER BY created_at ASC');
  return rows;
};

/**
 * Mendapatkan paper size berdasarkan ID
 * @param {number} id
 * @returns {Promise<Object|null>} Paper size object atau null
 */
export const getPaperSizeById = async (id) => {
  const [rows] = await db.query('SELECT * FROM paper_sizes WHERE id = ?', [id]);
  return rows.length > 0 ? rows[0] : null;
};

/**
 * Membuat paper size baru
 * @param {Object} data Data paper size 
 * @returns {Promise<number>} ID paper size baru
 */
export const createPaperSize = async (data) => {
  const query = 'INSERT INTO paper_sizes (name, top_margin, side_margin, vertical_pitch, horizontal_pitch, label_width, label_height, number_across, number_down, page_width, page_height) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)';
  const values = [
    data.name,
    data.top_margin || 0,
    data.side_margin || 0,
    data.vertical_pitch || 0,
    data.horizontal_pitch || 0,
    data.label_width || 0,
    data.label_height || 0,
    data.number_across || 1,
    data.number_down || 1,
    data.page_width || 0,
    data.page_height || 0
  ];
  const [result] = await db.query(query, values);
  return result.insertId;
};

/**
 * Memperbarui paper size
 * @param {number} id
 * @param {Object} data
 * @returns {Promise<boolean>} Status keberhasilan
 */
export const updatePaperSize = async (id, data) => {
  const query = 'UPDATE paper_sizes SET name = ?, top_margin = ?, side_margin = ?, vertical_pitch = ?, horizontal_pitch = ?, label_width = ?, label_height = ?, number_across = ?, number_down = ?, page_width = ?, page_height = ? WHERE id = ?';
  const values = [
    data.name,
    data.top_margin || 0,
    data.side_margin || 0,
    data.vertical_pitch || 0,
    data.horizontal_pitch || 0,
    data.label_width || 0,
    data.label_height || 0,
    data.number_across || 1,
    data.number_down || 1,
    data.page_width || 0,
    data.page_height || 0,
    id
  ];
  const [result] = await db.query(query, values);
  return result.affectedRows > 0;
};

/**
 * Menghapus paper size
 * @param {number} id
 * @returns {Promise<boolean>} Status keberhasilan
 */
export const deletePaperSize = async (id) => {
  const [result] = await db.query('DELETE FROM paper_sizes WHERE id = ?', [id]);
  return result.affectedRows > 0;
};

File: backend/internal/repository/paper_size_repository.go
package repository

import (
	"context"

	"github.com/dps-wmhris/backend/internal/dto"
	"github.com/jmoiron/sqlx"
)

type PaperSizeRepository interface {
	FindAll(ctx context.Context) ([]dto.PaperSizeResponse, error)
	FindByID(ctx context.Context, id int) (*dto.PaperSizeResponse, error)
	Insert(ctx context.Context, req dto.CreatePaperSizeRequest) (int, error)
	Update(ctx context.Context, id int, req dto.UpdatePaperSizeRequest) (int, error)
	Delete(ctx context.Context, id int) (int, error)
}

type paperSizeRepositoryImpl struct {
	db *sqlx.DB
}

func NewPaperSizeRepository(db *sqlx.DB) PaperSizeRepository {
	return &paperSizeRepositoryImpl{db: db}
}

func (r *paperSizeRepositoryImpl) FindAll(ctx context.Context) ([]dto.PaperSizeResponse, error) {
	query := "SELECT id, name, top_margin, side_margin, vertical_pitch, horizontal_pitch, label_width, label_height, number_across, number_down, page_width, page_height, is_active FROM paper_sizes ORDER BY id DESC"
	var sizes []dto.PaperSizeResponse
	err := r.db.SelectContext(ctx, &sizes, query)
	if err != nil {
		return nil, err
	}
	if sizes == nil {
		sizes = []dto.PaperSizeResponse{}
	}
	return sizes, nil
}

func (r *paperSizeRepositoryImpl) FindByID(ctx context.Context, id int) (*dto.PaperSizeResponse, error) {
	query := "SELECT id, name, top_margin, side_margin, vertical_pitch, horizontal_pitch, label_width, label_height, number_across, number_down, page_width, page_height, is_active FROM paper_sizes WHERE id = ?"
	var size dto.PaperSizeResponse
	err := r.db.GetContext(ctx, &size, query, id)
	if err != nil {
		return nil, err
	}
	return &size, nil
}

func (r *paperSizeRepositoryImpl) Insert(ctx context.Context, req dto.CreatePaperSizeRequest) (int, error) {
	query := `
		INSERT INTO paper_sizes (
			name, top_margin, side_margin, vertical_pitch, horizontal_pitch,
			label_width, label_height, number_across, number_down, page_width, page_height
		) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
	`
	res, err := r.db.ExecContext(ctx, query,
		req.Name, req.TopMargin, req.SideMargin, req.VerticalPitch, req.HorizontalPitch,
		req.LabelWidth, req.LabelHeight, req.NumberAcross, req.NumberDown, req.PageWidth, req.PageHeight,
	)
	if err != nil {
		return 0, err
	}
	id, err := res.LastInsertId()
	return int(id), err
}

func (r *paperSizeRepositoryImpl) Update(ctx context.Context, id int, req dto.UpdatePaperSizeRequest) (int, error) {
	query := `
		UPDATE paper_sizes SET
			name = ?, top_margin = ?, side_margin = ?, vertical_pitch = ?, horizontal_pitch = ?,
			label_width = ?, label_height = ?, number_across = ?, number_down = ?, page_width = ?, page_height = ?
		WHERE id = ?
	`
	res, err := r.db.ExecContext(ctx, query,
		req.Name, req.TopMargin, req.SideMargin, req.VerticalPitch, req.HorizontalPitch,
		req.LabelWidth, req.LabelHeight, req.NumberAcross, req.NumberDown, req.PageWidth, req.PageHeight,
		id,
	)
	if err != nil {
		return 0, err
	}
	affected, err := res.RowsAffected()
	return int(affected), err
}

func (r *paperSizeRepositoryImpl) Delete(ctx context.Context, id int) (int, error) {
	query := "DELETE FROM paper_sizes WHERE id = ?"
	res, err := r.db.ExecContext(ctx, query, id)
	if err != nil {
		return 0, err
	}
	affected, err := res.RowsAffected()
	return int(affected), err
}
